test/models/ethnic_church_builder_test.rb

```diff
@@ -0,0 +1,54 @@
+require 'test_helper'
+
+class EthnicChurchBuilderTest < ActiveSupport::TestCase
+  test 'EthnicChurchBuilder.instantiate' do
+    params = ActionController::Parameters.new({
+      ethnic_church: {
```

Comment:
Looking at your comment about this being a code smell. In my experience if you find that writing a test for a class is getting tedious then its doing too much. Another approach u can do here is dependency injection.  So above i mentioned that u could pass in ```
languages
country_name
address 
note_content ```
as the params. This would make these params in the test a lot smaller since u wouldnt have to worry about the structure of that hash. Dependency injection would make this even smaller by passing in the Language objects that would associate to the new EthnicChurch object. So your EthnicChurch class would take in
`(language_objects: language_objects, ...)`
where language_objects is a private method that could be defined in your controller which just gets the language objects similar to what u are already doing in the class itself.
```ruby
def languae_objects
  @params[:language][:name].reject(&:blank?).map do |lang_name|
    Language.find_or_initialize_by(name: lang_name)
  end
end
```

and in your EthnicChurchBuilder would just do ``` @ec.languages = language_objects``` Your model validations should worry about the name being empty.

you can do the same for country and  religious_background.

Reply:
I really like the idea of injecting all the necessary objects into `EthnicChurchBuilder`! I think the tradeoff is that I end up needing a ton of private methods in the controller pulling in data but as long as that isn't in the create action itself, that seems reasonable. I might even be able to slim some of those methods down a bit by offloading to the model.

Reply:
nothing wrong with private methods :) Here is another idea. Have a class that handles creating the language object. Call it LanguageBuilder that takes in the array of names.  There are many ways to go about this